### PR TITLE
Move Mock Exporter to Separate File

### PR DESCRIPTION
**Pull request by nholbrook**
Moved MockSpanExporter - a mock that was duplicated and being used in multiple files - to it's own file. The implementation of each of these mocks were slightly different, so I had to modify a few of the tests that use it. 

I decided to follow the existing folder structure of the existing exporters, so once completed this can easily be swapped out with an actual exporter.

### CODE REVIEW

File: exporters/mock/include/opentelemetry/exporters/mock/mock_span_exporter.h
Comment (pyohannes) [inline exporters/mock/include/opentelemetry/exporters/mock/mock_span_exporter.h:19]: Let's call this `InMemorySpanExporter` instead of `MockSpanExporter`.

Comment (pyohannes) [reply exporters/mock/include/opentelemetry/exporters/mock/mock_span_exporter.h:19]: Also, let's wrap `spans_received` and `shutdown_called` in a separate class (e. g. `InMemorySpanData`) that provides thread-safe access to the data. The current solution makes it impossible to access `spans_received` in a thread-safe manner.

Ideally, we would use the lock-free [CircularBuffer](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/src/common/circular_buffer.h), as it provides both thread safe access and the ability to manage a maximum capacity. An alternative implementation would be a vector with locks around each read and write.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (jsandler18) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:12]: nit: this is more of a fake then a mock
https://stackoverflow.com/questions/346372/whats-the-difference-between-faking-mocking-and-stubbing#:~:text=Fake%3A%20a%20class%20that%20implements,fixed%20data%20and%20no%20logic.&text=Mock%3A%20a%20class%20that%20implements,have%20been%20called%2Fnot%20called.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (oulin-coder) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:30]: This shouldn't need to be a shared_ptr. Bools take up negligible memory and can be passed by value, so using a shared_ptr just adds complexity.

File: exporters/memory/src/in_memory_span_exporter.cc
Comment (oulin-coder) [inline exporters/memory/src/in_memory_span_exporter.cc:27]: I left the same comment in Oliver's documentation PR, so copying here:

Hm, this takes ownership of the recordable object. This is fairly dangerous, since the param "recordables" will no longer have ownership of its Recordable members. Note that unique_ptr.release() is not a const method. I believe the compiler allows this because you're not mutating the input param "recordables", just the container objects (this is just my theory though).

The implementation right now makes Export take ownership over all recordables passed to it. Consider if this is the desired behavior for this exporter.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (oulin-coder) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:14]: Forgot to add this comment earlier, but implementations should be in a .cc file (except for templated classes/functions).

File: exporters/memory/src/in_memory_span_exporter.cc
Comment (nholbrook) [reply exporters/memory/src/in_memory_span_exporter.cc:27]: Addressed offline and was told to disregard this comment.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h]: Updated to use a normal bool and added a getter fucntion

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h]: Done

Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h]: Thanks - I've updated the terminology to reflect that this will be viewed as an in memory exporter as opposed to a mock.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:31]: Actually, this update caused some tests to fail. Adding a normal bool w/ a getter function won't work because the tests rely on SpanExporter (a parent class of all exporters), and the added getter function is not a standard function between all exporters.

Would you recommend I keep the current approach, add a public bool, or try a different approach?

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (oulin-coder) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:15]: Nope this was my bad. I was looking at another PR and didn't notice that this is just a class definition.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h]: Removed shutdown_called_ field, as it was only used for testing, as per Oulin's recommendation

File: exporters/mock/include/opentelemetry/exporters/mock/mock_span_exporter.h
Comment (nholbrook) [reply exporters/mock/include/opentelemetry/exporters/mock/mock_span_exporter.h]: Updated to InMemorySpanExporter and updated functionality to use CircularBuffer

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (pyohannes) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h:27]: Make `Add` and the constructor private and add the `InMemorySpanExporter` as a friend class.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (pyohannes) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:30]: I think it's better to return the complete `InMemorySpanData` from the getter. The exporter will be passed as an unique pointer to the processor and thus the user will not be able to call this getter.

File: exporters/memory/src/in_memory_span_data.cc
Comment (pyohannes) [inline exporters/memory/src/in_memory_span_data.cc:33]: I think you have to use `Swap` here. That's a bit cumbersome, we should add something to the interfaced of `AtomicUniquePtr` to make this more convenient. For now, do something similar to [here](https://github.com/open-telemetry/opentelemetry-cpp/pull/195/files#diff-8f953be5a9c81b02484d4faea832a5f6R189).

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h]: Updated

File: exporters/memory/src/in_memory_span_data.cc
Comment (nholbrook) [reply exporters/memory/src/in_memory_span_data.cc]: Updated

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h]: I'm not sure how to configure the configure the BUILD file to enable this. This would require adding the in_memory_span_data and in_memory_span_exporter as dependencies of each other, and would create a circular dependency. Any suggestions on how this can be done?

File: sdk/test/trace/sampler_benchmark.cc
Comment (nholbrook) [inline sdk/test/trace/sampler_benchmark.cc:1]: @pyohannes This test is still unable to find my exporter. Could you take a look at my build and cmake file and see if there is something misconfigured? I'm using the existing exporters as reference, but you have more experience with this setup and might be able to see something I'm overlooking.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (pyohannes) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h:12]: Could we move all unused headers into the `*.cc` file? Like `atomic_unique_ptr.h`, `circular_buffer_range.h` and `span_data.h`.

Also, let's avoid `using` directives in the header file. Everybody including the header file will get those, which can cause namespace pollution.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (pyohannes) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:30]: After thinking a bit more about this, let's make this return a `std::shared_ptr<InMemorySpanData>`, and let's store it as a shared pointer internally.

Otherwise, one cannot safely hold on to the object returned, as it might be invalidated once the exporter is destroyed.

File: exporters/memory/src/in_memory_span_data.cc
Comment (pyohannes) [inline exporters/memory/src/in_memory_span_data.cc:14]: Let's add a constructor argument for the maximum buffer size, using MAX_BUFFER_SIZE as a default.

File: exporters/memory/src/in_memory_span_data.cc
Comment (pyohannes) [inline exporters/memory/src/in_memory_span_data.cc:18]: This file is not correctly formatted.

File: exporters/memory/src/in_memory_span_exporter.cc
Comment (pyohannes) [inline exporters/memory/src/in_memory_span_exporter.cc:10]: Let's add an (optional) argument for the `InMemorySpanData` buffer size.

File: sdk/test/trace/sampler_benchmark.cc
Comment (pyohannes) [reply sdk/test/trace/sampler_benchmark.cc:1]: This isn't very elegant at all, but let's try a relative include here:

```c
 #include "../../../exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h"
```

I think the problem you're seeing is related to a circular dependency (the sdk and its tests depend on the memory exporter, and the memory exporter in turn depends on the SDK). If the relative import doesn't work, we might have to change things in the global `CMakeLists.txt`.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h]: Done

File: exporters/memory/src/in_memory_span_data.cc
Comment (nholbrook) [reply exporters/memory/src/in_memory_span_data.cc]: Done

File: exporters/memory/src/in_memory_span_exporter.cc
Comment (nholbrook) [reply exporters/memory/src/in_memory_span_exporter.cc]: Done

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h]: Done

File: exporters/memory/src/in_memory_span_data.cc
Comment (nholbrook) [reply exporters/memory/src/in_memory_span_data.cc]: Formatted

File: sdk/test/trace/sampler_benchmark.cc
Comment (nholbrook) [reply sdk/test/trace/sampler_benchmark.cc]: This workaround works for `./ci/run_docker.sh ./ci/do_ci.sh cmake.test`, but fails for bazel. When running one test vs another, it's requiring a separate set of imports. See below.

For tests:
```
// Import for CMake
// #include "../../../exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h"

// Import for Bazel
#include "opentelemetry/exporters/memory/in_memory_span_exporter.h"
```

And similarly for the exporter itself:
```
// Import for CMake
// #include "in_memory_span_data.h"

// Import for Bazel
#include "opentelemetry/exporters/memory/in_memory_span_data.h"
```

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (pyohannes) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h:19]: I think one of the easiest solutions to make this work would be to make this exporter a header-only implementation.

Could you move all the implementation into the header files? Then circular dependencies shouldn't be a problem.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (pyohannes) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:5]: This should work for both CMake and Bazel.

https://gcc.gnu.org/onlinedocs/cpp/Search-Path.html

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h]: Done

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h]: The changes I made when I moved to a header-only implementation fixed that import issue, I just forgot to remove the comments. I've went ahead and fixed that

File: sdk/test/trace/sampler_benchmark.cc
Comment (nholbrook) [reply sdk/test/trace/sampler_benchmark.cc:1]: Resolved

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (oulin-coder) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h:42]: Add a comment on why we're doing this pointer swap? It's not immediately clear to the reader that it's a requirement of the Consume function that the AtomicUniquePtr must be set to null (I had to go and read about it).

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (oulin-coder) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:25]: nit: it's more readable to always add the parenthesis when calling constructors/methods.

File: exporters/memory/test/in_memory_span_data_test.cc
Comment (oulin-coder) [inline exporters/memory/test/in_memory_span_data_test.cc:18]: You should be able to just do std::make_unique\<SpanData\>(). Since Recordable is a parent of SpanData, the cast should be implicit.

File: exporters/memory/test/in_memory_span_data_test.cc
Comment (oulin-coder) [inline exporters/memory/test/in_memory_span_data_test.cc:17]: I think you can just define this as a span of SpanData instead of Recordable, since you just convert it back to SpanData later.

File: exporters/memory/test/in_memory_span_data_test.cc
Comment (oulin-coder) [inline exporters/memory/test/in_memory_span_data_test.cc:31]: nit: this comment is for the first ASSERT right?

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (pyohannes) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h:1]: I advised @nholbrook to do this. It is to avoid circular binary dependencies between the SDK and the in-memory exporter.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h:22]: Added

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h:55]: Added

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h]: Added parenthesis

File: exporters/memory/test/in_memory_span_data_test.cc
Comment (nholbrook) [reply exporters/memory/test/in_memory_span_data_test.cc]: make_unique was added in C++ 14, and we're running on C++ 11, so this won't be possible

File: exporters/memory/test/in_memory_span_data_test.cc
Comment (nholbrook) [reply exporters/memory/test/in_memory_span_data_test.cc]: Updated

File: exporters/memory/test/in_memory_span_data_test.cc
Comment (nholbrook) [reply exporters/memory/test/in_memory_span_data_test.cc]: Correct, I've went ahead and moved it to clear up any confusion

File: exporters/memory/test/in_memory_span_exporter_test.cc
Comment (pyohannes) [inline exporters/memory/test/in_memory_span_exporter_test.cc:18]: This should rather be something along this lines:
```c++
std::unique_ptr<Recordable> spandata(new SpanData());
opentelemetry::nostd::span<std::unique_ptr<Recordable>> batch(&spandata, 1);
```
Otherwise we'll leak the memory of the dynamically allocated `unique_ptr` (`new std::unique_ptr<Recordable>`).

File: sdk/test/trace/simple_processor_test.cc
Comment (pyohannes) [inline sdk/test/trace/simple_processor_test.cc:2]: I assume we don't need this include, as it is included by `in_memory_span_exporter.h`.

File: exporters/memory/test/in_memory_span_data_test.cc
Comment (pyohannes) [inline exporters/memory/test/in_memory_span_data_test.cc:18]: Same here.

File: exporters/memory/test/in_memory_span_data_test.cc
Comment (pyohannes) [inline exporters/memory/test/in_memory_span_data_test.cc:20]: Do we need this loop, when the above span only contains one recordable?

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (pyohannes) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:56]: Let's add a short note that this is not supported for this exporter.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (pyohannes) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:48]: You can omit the `get()` here:
```c++
data_->Add(std::move(span));
```

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (pyohannes) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:5]: Let's move this inside the namespace. Otherwise we pollute the global namespace.

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h]: Removed

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h:56]: Added

File: exporters/memory/test/in_memory_span_data_test.cc
Comment (nholbrook) [reply exporters/memory/test/in_memory_span_data_test.cc]: Nope, I've removed that

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_exporter.h]: Ok, done

File: exporters/memory/test/in_memory_span_exporter_test.cc
Comment (nholbrook) [reply exporters/memory/test/in_memory_span_exporter_test.cc]: Alright, updated

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (nholbrook) [reply exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h]: Updated

**Comment by nholbrook** [comment]
@pyohannes Is there anything else that you see that needs to be addressed before this can be merged? This is my last day and if this doesn't get approved today I'll have to close this PR and reopen a new one next week due to CLA issues with my google email being deleted

**Comment by pyohannes** [comment]
> Is there anything else that you see that needs to be addressed before this can be merged? 

I see the Bazel tests failing, it seems that's related to the context work. I'll have a look at that.

**Comment by pyohannes** [comment]
@open-telemetry/cpp-approvers This is ready for review.

The code coverage test fails for reasons I cannot really understand.

**Review by reyang — APPROVED** [review]

File: exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h
Comment (ThomsonTan) [inline exporters/memory/include/opentelemetry/exporters/memory/in_memory_span_data.h:6]: For header file from standard library, prefer including it via `#include <vector>`?

**Pull request merged**
